File: shoebill_plugins/LtrpApi/src/main/java/lt/ltrp/ItemController.java
package lt.ltrp;

import lt.ltrp.constant.ItemType;
import lt.ltrp.dao.ItemDao;
import lt.ltrp.dao.PhoneDao;
import lt.ltrp.object.*;
import net.gtaun.shoebill.constant.SpecialAction;
import net.gtaun.util.event.EventManager;

/**
 * @author Bebras
 *         2016.04.14.
 */
public interface ItemController {

    WeedSeedItem createWeedSeed(EventManager eventManager);

    WeedItem createWeed(EventManager eventManager, int doses);



    class Instance {
        static ItemController instance;
    }

    static ItemController get() {
        return Instance.instance;
    }

    Inventory createInventory(EventManager eventManager, InventoryEntity owner, String name, int size);
    Item createItem(ItemType type, InventoryEntity owner, EventManager eventManager);
    Item createItem(ItemType type, String name, InventoryEntity entity, EventManager eventManager);
    Item createItem(ItemType type, String name, SpecialAction specialAction, InventoryEntity entity, EventManager eventManager);

    ItemDao getItemDao();
    PhoneDao getPhoneDao();

}

File: shoebill_plugins/ItemPlugin/src/main/java/lt/ltrp/item/ItemController.kt
package lt.ltrp.item

import lt.ltrp.`object`.*
import lt.ltrp.constant.ItemType
import net.gtaun.shoebill.constant.SpecialAction
import net.gtaun.util.event.EventManager

/**
 * Created by Bebras on 2016-11-09.
 * This class exists only for legacy reasons
 */
@Deprecated("Don't use, use ItemFactory and InventoryFactory instead")
interface ItemController {

    fun createInventory(eventManager: EventManager, owner: InventoryEntity, name: String, size: Int): Inventory

    fun createItem(type: ItemType, name: String, entity: InventoryEntity, eventManager: EventManager): Item
    fun createItem(type: ItemType, owner: InventoryEntity, eventManager: EventManager): Item
    fun createItem(type: ItemType, name: String, specialAction: SpecialAction, owner: InventoryEntity, eventManager: EventManager): Item

    companion object {
        private var instance = ItemControllerImpl()
        fun get(): ItemController {
            return instance
        }
    }
}
private class ItemControllerImpl() : ItemController {

    override fun createInventory(eventManager: EventManager, owner: InventoryEntity, name: String, size: Int): Inventory {
        return InventoryFactory.INSTANCE.create(owner, name, size)
    }

    override fun createItem(type: ItemType, name: String, entity: InventoryEntity, eventManager: EventManager): Item {
        return ItemFactory.instance.create(type, name, entity, eventManager)
    }

    override fun createItem(type: ItemType, owner: InventoryEntity, eventManager: EventManager): Item {
        return ItemFactory.instance.create(type, owner, eventManager)
    }

    override fun createItem(type: ItemType, name: String, specialAction: SpecialAction, owner: InventoryEntity, eventManager: EventManager): Item {
        return ItemFactory.instance.create(type, name, specialAction, owner, eventManager)
    }

}